inc/gpio.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+// GPIO HAL Interface
+
+#include <stdbool.h>
+#include <stdint.h>
+
+// IO Pin "address" to be used to change that pin's settings
+// Use bytes instead of words to save on memory also convenient for MSP430
+// compatibility
+typedef struct IOAddr {
+  const uint8_t IOPort;  // Could be an enum (A-F) for stm32f0xx
+  const uint8_t IOPin;
+} IOAddr;
+
+// For setting the direction of the pin
+typedef enum { PIN_IN = 0; PIN_OUT; }
+PIN_DIR;
+
+// For setting the output value of the pin
+typedef enum { PIN_LOW = 0; PIN_HIGH; }
+PIN_VAL;
+
+// For setting the internal pull-up/pull-down resistor (Turn off if a hardware
+// PUPDR is enabled)
+// Different on MSP and STM for MSP RES_UP and RES_DOWN both set PIN_REN to 1
+// Could make this RES_ON, RES_OFF and make pullup and pulldown vary with the
+// PIN_VAL as it works
+// for the MSP430
+typedef enum { RES_OFF = 0; RES_UP; RES_DOWN; }
+PIN_RES;
+
+// For setting the alternate function on the pin
+// Applicable to STM, on MSP just set the PIN_SEL to 1 unless AF0 is passed
+// On STM if an AF is passed set PIN_MODER to 2 instead of IN or OUT, ANALOG
+// sets to 3
+typedef enum {
+  AF0 = 0; AF1; AF2; AF3; AF4; AF5; AF6; AF7;
+  ANALOG;  // Clever way to make DIR analog for STM as AF and Analog are
+           // mutually exclusive
+}
+PIN_AF;
+
+// Ignored settings:
+// Speed - should be set when a pin is initialized and not touched again
+// Output type - very rarely will we want open drain, unless we have shared
+// input or wired OR gates
+// Reset and Lock setting on STM
+
+// A struct to set the values for a pin
+typedef struct IOSettings {
+  const IOAddr addr;  // The necessity of making this const is debatable
+  PIN_DIR dir;
+  PIN_VAL val;
+  PIN_RES res;
+  PIN_AF alt;
+} IOSettings;
+
+// Initializes a pin with all of the settings entered
+void gpio_init(IOSettings *settings);
```

Comment:
It might be better to gpio_init be a global gpio driver init and something like that gpio_init_pin initialize individual pins.